streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java

```diff
@@ -114,7 +103,12 @@ int addRawRecords(final TopicPartition partition, final Iterable<ConsumerRecord<
         final RecordQueue recordQueue = partitionQueues.get(partition);
 
         final int oldSize = recordQueue.size();
+        final long oldTimestamp = recordQueue.timestamp();
         final int newSize = recordQueue.addRawRecords(rawRecords);
+        final long newTimestamp = recordQueue.timestamp();
+        if (newTimestamp != oldTimestamp && newTimestamp > streamTime) {
```

Comment:
With current implementation using all buffered record to compute "stream time" I think we need to recompute "stream time" every time we add new records, don't we? Maybe we can improve when we switch to considering "head records" only?

Reply:
Yes, we basically are recomputing it every time we add new records right now (if adding the records changed the queue time). I think we can use the same optimization discussed above.

I think that considering head-records only, you can optimize this case, because then, you can only change the queue time by adding records when the queue is empty. Whereas now, you can change the queue time any time you add new records (because you might add a new min).